tests/ext/drv/test_ext_drv.c

```diff
@@ -0,0 +1,298 @@
+        // after WdfDeviceCreate fails.
+        WdfDeviceInitFree(device_initialize);
+        device_initialize = NULL;
+        goto Exit;
+    }
+
+    device_create_flag = TRUE;
+
+    // Create symbolic link for control object for user mode.
+    RtlInitUnicodeString(&test_ebpf_ext_symbolic_device_name, TEST_EBPF_EXT_SYMBOLIC_DEVICE_NAME);
+    status = WdfDeviceCreateSymbolicLink(*device, &test_ebpf_ext_symbolic_device_name);
+
+    if (!NT_SUCCESS(status)) {
+        goto Exit;
+    }
+
+    // parallel default queue
+    WDF_IO_QUEUE_CONFIG_INIT_DEFAULT_QUEUE(&io_queue_configuration, WdfIoQueueDispatchParallel);
+
+    io_queue_configuration.EvtIoDeviceControl = _test_ebpf_ext_driver_io_device_control;
+
+    status = WdfIoQueueCreate(
+        *device,
+        &io_queue_configuration,
+        WDF_NO_OBJECT_ATTRIBUTES,
+        WDF_NO_HANDLE // pointer to default queue
+    );
+    if (!NT_SUCCESS(status)) {
+        goto Exit;
+    }
+
+    status = test_ebpf_extension_program_info_provider_register();
+    if (!NT_SUCCESS(status)) {
+        goto Exit;
+    }
+
+    status = test_ebpf_extension_hook_provider_register();
+    if (!NT_SUCCESS(status)) {
+        goto Exit;
+    }
+
+    WdfControlFinishInitializing(*device);
+
+Exit:
+    if (!NT_SUCCESS(status)) {
+        if (device_create_flag && device != NULL) {
+            //
+            // Release the reference on the newly created object, since
+            // we couldn't initialize it.
+            //
+            WdfObjectDelete(*device);
+        }
+    }
+    return status;
+}
+
+NTSTATUS
+DriverEntry(_In_ DRIVER_OBJECT* driver_object, _In_ UNICODE_STRING* registry_path)
+{
+    NTSTATUS status;
+    WDFDRIVER driver;
+    WDFDEVICE device;
+
+    // Request NX Non-Paged Pool when available
+    ExInitializeDriverRuntime(DrvRtPoolNxOptIn);
+
+    KdPrintEx((DPFLTR_IHVDRIVER_ID, DPFLTR_INFO_LEVEL, "test_ebpf_ext: DriverEntry\n"));
+
+    status = _test_ebpf_ext_driver_initialize_objects(driver_object, registry_path, &driver, &device);
+
+    if (!NT_SUCCESS(status)) {
+        goto Exit;
+    }
+
+    _test_ebpf_ext_driver_device_object = WdfDeviceWdmGetDeviceObject(device);
+
+Exit:
+
+    return status;
+}
+
+static VOID
+_test_ebpf_ext_driver_io_device_control(
+    _In_ WDFQUEUE queue,
+    _In_ WDFREQUEST request,
+    size_t output_buffer_length,
+    size_t input_buffer_length,
+    ULONG io_control_code)
+{
+    NTSTATUS status = STATUS_SUCCESS;
+    WDFDEVICE device;
+    void* input_buffer = NULL;
+    void* output_buffer = NULL;
+    size_t actual_input_length = 0;
+    size_t actual_output_length = 0;
+    test_program_context_t program_context = {0};
+    uint32_t program_result = 0;
+    ebpf_result_t result;
+
+    device = WdfIoQueueGetDevice(queue);
+
+    switch (io_control_code) {
+    case IOCTL_TEST_EBPF_EXT_CTL:
+        if (input_buffer_length != 0) {
+            // Retrieve the input buffer associated with the request object
```

Comment:
```suggestion
            // Retrieve the input buffer associated with the request object.
```
(nit: most comments end in punctuation, so tagging a couple more for consistency)